sklearn/cluster/spectral.py

```diff
@@ -263,7 +263,7 @@ def spectral_clustering(affinity, n_clusters=8, n_components=None,
     maps = spectral_embedding(affinity, n_components=n_components,
                               eigen_solver=eigen_solver,
                               random_state=random_state,
-                              eigen_tol=eigen_tol, drop_first=False)
+                              eigen_tol=eigen_tol, drop_first=True)
```

Comment:
Not a SpectralClustering expert at all, but according to the comment above, there seems to be a reason for `drop_first=False`:

What do you make of this comment?
```py
# The first eigen vector is constant only for fully connected graphs 
# and should be kept for spectral clustering (drop_first = False) 
# See spectral_embedding documentation. 
```

Reply:
I have no idea about why the comment said drop_first should false. 

All information about SpectralClustering I read show that the eigenvector that correspond to eigenvalues 0 should be drop, which is the first eigenvector. And it's also reasonable in the physics view. 

Take information in wiki (https://en.wikipedia.org/wiki/Spectral_clustering) for example: 
**The eigenvectors that are relevant are the ones that correspond to smallest several eigenvalues of the Laplacian except for the smallest eigenvalue which will have a value of 0**

Anyway, my test example given above is simple and obvious.

Reply:
An obvious reason to drop the first eigenvector is that the elements in this eigenvector are identical, for example like [4, 4, 4, 4, 4, 4]. As I have said in the issue, it represent a translational motion of the system in physics. It give no information about how to devide nodes into clusters. 

@jmargeta may know well this property of this eigenvector according to https://github.com/scikit-learn/scikit-learn/pull/9062

In contrast, the eigenvector related to the second smallest eigenvalue in my example is something like [4, 4, 3, -4, -4, -3].  Nodes in different clusters have different sign. Not dropping the first eigenvector only affects the stability of the K-means result.

Reply:
I agree with @sky88088 that **for fully connected graphs**, the first eigenvector is constant and **should be dropped** as it actually hurt clustering performance when included.
However, when the graph contains disconnected regions, isn't it precisely the first - best splitting - eigenvector that you would want to keep?

Most of the cases have fully connected graphs and if not, getting the clusters spectrally is probably not even needed and probably should be the default.

Extra graphical intuition here: https://www.cs.cmu.edu/~aarti/Class/10701/slides/Lecture21_2.pdf

Reply:
I do agree with @jmargeta. 

However, even in the case of graph contains disconnected region, I think one can get the right result only when setting the exact right n_components=the number of disconnected region.  In this case, it's hard for users to make the right choise in practice. If I'm wrong, correct me plz. 

Then the first eigenvector hurt the performance of clustering in the most of case, and it's hard to use even when it's useful.

Maybe we should check the connectivity of the input matrix before applying spectral_embedding to decide whether drop the first eigenvector. Or just raise an error for this case and drop the first eigenvector in other cases.

Reply:
Scikit-learn is a wonderful tool also because makes many of the hard choices for its users out of the box. Getting this right is important and thanks for looking into this @sky88088 .

I do not know whether there is any consensus on what to do with disconnected regions. 
- If we set n_components=number of disconnected regions, would there be any need for clustering? (what about assigning a unique label to each connected component?)
- What about isolated outliers? Do we count them as valid connected components?
- ...

Dropping the first should only hurt the performance with disconnected graphs. So a connectivity test to decide whether to drop or not seems to me like a good step forward.

Reply:
I think we now have the common idea that the first eigenvector should be dropt in the most of cases. 

The second improvement is a little complicated. I don't know if there is any variant of spectral clustering or spectral embedding use eigenvalues to weight their eigenvectors. I learn this idea from the elastic network models(https://en.wikipedia.org/wiki/Gaussian_network_model) in physics. Actually, it's almost the same thing of spectral embedding. 

Whether to drop the first eigenvector now is a parameter of the function. I don't know if I should add a new parameter to decide whether to weigh the eigenvectors. In my view, this will almost always improve the result of embedding, so there is no need for users to choose do it or not. However, not like classification task, it maybe not easy to judge which embedding result is better. 

In my example, this improvement make the result robust and independent of the choise of the number of components. But my example is simple and maybe not so convincing.

Reply:
Reproducible research is something important and I haven't taken it into account. Then adding a new parameter to decide whether to introduce weighting I think is necessary.